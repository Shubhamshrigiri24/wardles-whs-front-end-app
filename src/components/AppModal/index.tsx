import React, { FC, useState, useCallback } from "react";
import { Grid } from "@material-ui/core";
import ChevronRight from "@material-ui/icons/ChevronRight";
import {
  Typography,
  Modal,
  Button,
  Spacing,
  wellColors,
} from "@welldigital/components";
import { useStyles } from "./styles";

export type AppModalProps = {
  isOpen: boolean;
  title?: string;
  backLabel?: string;
  successLabel?: string;
  onSuccess: () => void;
  onBack: () => void;
};

const AppModal: FC<AppModalProps> = ({
  isOpen,
  title = "",
  backLabel = "Back",
  successLabel = "Continue",
  children,
  onSuccess,
  onBack,
}) => {
  const classes = useStyles();
  const [isBusy, setIsBusy] = useState<boolean>(false);

  const onInternalSuccess = useCallback(async () => {
    setIsBusy(true);
    await onSuccess();
    setIsBusy(false);
  }, [onSuccess]);

  const onInternalBack = useCallback(async () => {
    setIsBusy(true);
    await onBack();
    setIsBusy(false);
  }, [onBack]);

  return isOpen ? (
    <Modal
      width={576}
      open={isOpen}
      onClose={onInternalBack}
      maxWidth={"xs"}
      PaperProps={{ classes: { root: classes.paper } }}
    >
      <Spacing spacing={2} />
      <Typography
        variant={"h3"}
        align={"center"}
        spacingAfter={2}
        className={classes.title}
        data-testid={"appModal/title"}
      >
        {title}
      </Typography>
      <div className={classes.body}>{children}</div>
      <Grid container justify={"center"} alignItems={"center"} spacing={2}>
        <Grid item>
          <Button
            onClick={onInternalBack}
            disabled={isBusy}
            variant={"text"}
            color={"default"}
            className={classes.backButton}
            data-testid={"appModal/backButton"}
          >
            <Typography style={{ color: wellColors.elixir[900] }}>
              {backLabel}
            </Typography>
          </Button>
        </Grid>
        <Grid item>
          <Button
            onClick={onInternalSuccess}
            loading={isBusy}
            variant={"contained"}
            color={"primary"}
            endIcon={<ChevronRight />}
            data-testid={"appModal/successButton"}
          >
            {successLabel}
          </Button>
        </Grid>
      </Grid>
    </Modal>
  ) : null;
};

export default AppModal;
