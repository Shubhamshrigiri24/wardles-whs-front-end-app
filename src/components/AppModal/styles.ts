import { makeStyles, Theme } from "@material-ui/core";
import { wellColors } from "@welldigital/components";

export const useStyles = makeStyles((theme: Theme) => ({
  title: {
    color: wellColors.elixir[900],
  },

  body: {
    color: wellColors.elixir[300],
  },

  paper: {
    height: "auto !important",
    maxHeight: "100%",
  },

  backButton: {
    minWidth: 60,
  },
}));
