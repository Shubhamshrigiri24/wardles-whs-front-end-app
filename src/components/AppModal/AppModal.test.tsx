import React from "react";
import { render, fireEvent, act } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import AppModal, { AppModalProps } from "./index";

const mockOnBack = jest.fn();
const mockOnSuccess = jest.fn();

beforeEach(() => {
  jest.resetAllMocks();
});

const makeProps = (props?: Partial<AppModalProps>): AppModalProps => {
  return {
    isOpen: true,
    onBack: mockOnBack,
    onSuccess: mockOnSuccess,
    ...props,
  };
};

describe("<AppModal />", () => {
  it("should match snapshot", () => {
    const props = makeProps();
    const { baseElement } = render(<AppModal {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  it("should open", async () => {
    const props = makeProps({ title: "Test modal title" });
    const { getByTestId } = render(<AppModal {...props} />);
    expect(getByTestId("appModal/title")).toBeInTheDocument();
    expect(getByTestId("appModal/title")).toHaveTextContent("Test modal title");
  });

  it("should fire Back button", async () => {
    const props = makeProps();
    const { getByTestId } = render(<AppModal {...props} />);
    const backButton = getByTestId("appModal/backButton");
    await act(async () => {
      fireEvent.click(backButton);
    });
    expect(mockOnBack).toHaveBeenCalledTimes(1);
  });

  it("should fire Success button", async () => {
    const props = makeProps();
    const { getByTestId } = render(<AppModal {...props} />);
    const successButton = getByTestId("appModal/successButton");
    await act(async () => {
      fireEvent.click(successButton);
    });
    expect(mockOnSuccess).toHaveBeenCalledTimes(1);
  });

  it("should render children", () => {
    const props = makeProps();
    const { getByTestId } = render(
      <AppModal {...props}>
        <div data-testid={"appModal/children"}>test children</div>
      </AppModal>
    );
    expect(getByTestId("appModal/children")).toBeInTheDocument();
  });

  it("should render default label for buttons", () => {
    const props = makeProps();
    const { getByText } = render(<AppModal {...props} />);
    ["Back", "Continue"].forEach((label) => {
      expect(getByText(label)).toBeInTheDocument();
    });
  });

  it("should pass label to back button", () => {
    const props = makeProps({ backLabel: "Go back" });
    const { getByText } = render(<AppModal {...props} />);
    expect(getByText("Go back")).toBeInTheDocument();
  });

  it("should pass label to success button", () => {
    const props = makeProps({ backLabel: "Next" });
    const { getByText } = render(<AppModal {...props} />);
    expect(getByText("Next")).toBeInTheDocument();
  });
});
