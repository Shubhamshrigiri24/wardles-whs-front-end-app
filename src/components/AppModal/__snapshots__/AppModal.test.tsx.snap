// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppModal /> should match snapshot 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root makeStyles-root-5"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root makeStyles-paper-3 MuiDialog-paper makeStyles-paper-7 MuiDialog-paperScrollPaper MuiDialog-paperWidthFalse MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
        style="width: 576px;"
      >
        <div
          class="MuiBox-root MuiBox-root-8"
          data-testid="modal/content"
        >
          <div
            class="makeStyles-spacing2-13"
            data-testid="spacing/root"
          />
          <h3
            class="MuiTypography-root makeStyles-title-1 MuiTypography-h3 MuiTypography-alignCenter"
            data-testid="appModal/title"
          />
          <div
            class="makeStyles-spacing2-13 makeStyles-spacing2Responsive-14"
            data-testid="spacing/root"
          />
          <div
            class="makeStyles-body-2"
          />
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <button
                class="MuiButtonBase-root MuiButton-root makeStyles-backButton-4 MuiButton-text makeStyles-backButton-4"
                data-testid="appModal/backButton"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  <div
                    class="makeStyles-children-37"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1"
                      style="color: rgb(12, 22, 31);"
                    >
                      Back
                    </p>
                  </div>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                data-testid="appModal/successButton"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  <div
                    class="makeStyles-children-37"
                  >
                    Continue
                  </div>
                  <span
                    class="MuiButton-endIcon makeStyles-endIcon-39 MuiButton-iconSizeMedium"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                      />
                    </svg>
                  </span>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiIconButton-root makeStyles-closeButton-6"
          data-testid="modal/close-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
